ide/coqide/protocol/serialize.ml

```diff
@@ -110,16 +110,44 @@ let to_edit_id = function
   | x -> raise (Marshal_error("edit_id",x))
 
 let of_loc loc =
-  let start, stop = Loc.unloc loc in
-  Element ("loc",[("start",string_of_int start);("stop",string_of_int stop)],[])
+  let {
+    Loc.fname
+  ; line_nb
+  ; bol_pos
+  ; line_nb_last
+  ; bol_pos_last
+  ; bp
+  ; ep
+  } = loc in
+  Element ("loc",
+           [ ("line_nb", string_of_int line_nb)
+           ; ("bol_pos", string_of_int bol_pos)
+           ; ("line_nb_last", string_of_int line_nb_last)
+           ; ("bol_pos_last", string_of_int bol_pos_last)
+           ; ("start", string_of_int bp)
+           ; ("stop", string_of_int ep )
+           ],
```

Comment:
Returning values other than `start` and `stop` isn't helpful.  For Pierre-Marie's example, the returned `line_nb` may be incorrect (and unusable) because new line(s) were inserted in the buffer above the sentence.  Same thing for `bol_pos`, `line_nb_last` and `bol_pos_last`.  All these values could be incorrect and would need to be recomputed in CoqIDE, so there's no point in sending them in the first place.

Reply:
Your argument seems strange as by it, also `start` and `stop` would be useless as they would have to be recomputed in Pierre-Marie's example.

Anyways they are recomputed fine and things work OK, and moreover the current code makes sense from an implementation and data ownership properties.

Line/column location is actually pretty useful due to the way the GTK API work (and other editors can for sure profit from it). The benefits are already clear if you look at the code in this PR vs the code in 8.16.

Reply:
How would you return a correct value for `bol_pos` for stmt2 for the following input?  The two statements are passed in separate calls to the parser (`(* comment *)\nstmt1.` and ` stmt2.`).  Would you have the parser retain this state or require that CoqIDE always pass the previous `bol_pos` (which is not easy to manage)?  Same thing for `bol_pos_last`.

It's not worth the trouble to make this work (likely fragile in CoqIDE) and it's silly to return values that the GUI can't rely on.

```
(* comment *)
stmt1. stmt2.
```

In `idetop.ml`, the earlier code was like this.  There are two problems.  First, the parser ignores `bol_pos`.  Second, it's not convenient for CoqIDE to pass the correct value.  Hence my comment "this won't yield correct values for bol_pos_last".

```
let add ((((s,eid),(sid,verbose)),off),(line_nb,bol_pos)) =
  let doc = get_doc () in
  let open Loc in
  (* note: this won't yield correct values for bol_pos_last,
     but the debugger doesn't use that *)
  let loc = { (initial ToplevelInput) with bp=off; line_nb; bol_pos } in   (* bol_pos IGNORED *)
  let r_stream = Gramlib.Stream.of_string ~offset:off s in
  let pa = Pcoq.Parsable.make ~loc r_stream in
```

Reply:
Your branch has exactly the code shown above, so I very much doubt it.  Try adding some print statements in the lexer to see what is returned for my little example above.

Reply:
IINM this test suggests that it's incorrect:  (I made the tactic that gets the warning the second sentence on the line.)

![image](https://user-images.githubusercontent.com/1253341/226477138-47b77b8b-d7d3-41ce-b2f9-68c4ee37e2ca.png)

Reply:
I'll have a look thanks, but if we can set `bp` correctly there is certainly not a problem setting `bol_pos` correctly too.

Reply:
Even if you can, it adds a bunch more bookkeeping (and adjustment) on the CoqIDE side.

Reply:
It adds maybe one line of bookeeping, true, but on the other it provides accurate info on line / column which is very useful for other endeavors, like for example reusing GTK's index-to-char UTF conversion methods, so the tradeoff is a clear win.

Moreover quite a few parts of the interface could use accurate line/column info.

Reply:
There were indeed two bugs here, see commit messages.

Reply:
Actually the bug was pretty subtle if you are curious, we were setting `bol_pos` correctly, but not `bol_pos_last`, which turns out to be used by the lexer in this particular example so the location with the warning the lexer used `bol_pos_last` setting bol_pos to the incorrect value.

So far all test cases look good to me.